sirepo/auth/ldap.py

```diff
@@ -0,0 +1,87 @@
+"""LDAP authentication support
+:copyright: Copyright (c) 2018-2022 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from pykern import pkconfig
+from pykern import pkinspect
+from pykern.pkdebug import pkdlog, pkdp
+from pykern.pkcollections import PKDict
+import ldap
+import re
+import sirepo.quest
+import sirepo.util
+
+# TODO(BG): Write test
+
+AUTH_METHOD = "ldap"
+
+AUTH_METHOD_VISIBLE = True
+
+_cfg = None
+
+_ESCAPE_DN_MAIL = re.compile(r"([,\\#<>;\"=+])")
+
+#: module handle
+this_module = pkinspect.this_module()
+
+#: Well known alias for auth
+user_model = "AuthEmailUser"
+
+
+class API(sirepo.quest.API):
+    @sirepo.quest.Spec("require_cookie_sentinel")
+    def api_authLdapLogin(self):
+
+        # validates username/password, escapes special chars for ldap dn
+        def _validate_escape_credentials(req):
+            email = req.req_data.email.lower()
+            if len(email) > 256 or len(req.req_data.password) > 256:
```

Comment:
This is actually more complicated than 256 ([really 255](https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc756101(v=ws.10)?redirectedfrom=MSDN#name-length-limitations-for-ldap-simple-bind-operations)). You will be adding base_dn. Moreover, the AD limits are a bit strange, and I think passwords [are limited to 127 characters](https://learn.microsoft.com/en-us/previous-versions/tn-archive/cc512606(v=technet.10)?redirectedfrom=MSDN) due to limitations in the Windows login dialogue. Technically, LDAP itself has no limits.

I would be more conservative here just because you don't want to deal with weirdnesses in the server, e.g. truncating passwords before hashing. Define a constant of MAX_ENTRY to be 127, and use that everywhere. 

You can check base_dn against this number as well and thrown an error in the config code